File: 05_valid_palindrome.js
/*
A phrase is a palindrome if, after converting all uppercase letters into 
lowercase letters and removing all non-alphanumeric characters, it reads 
the same forward and backward. 

Alphanumeric characters include letters and numbers.

Given a string s, return true if it is a palindrome, or false otherwise.
*/

// two pointers + regex
var isPalindrome = function (s) {
  // remove non-alphanumeric characters and convert to lowercase
  s = s.toLowerCase().replace(/[^a-z0-9]/g, "");
  let left = 0,
    right = s.length - 1;
  while (left < right) {
    if (s[left] !== s[right]) return false;
    left++;
    right--;
  }
  return true;
};

File: 05_valid_palindrome.py
# A phrase is a palindrome if, after converting all uppercase letters into 
# lowercase letters and removing all non-alphanumeric characters, it reads 
# the same forward and backward. 

# Alphanumeric characters include letters and numbers.

# Given a string s, return true if it is a palindrome, or false otherwise.

# https://leetcode.com/problems/valid-palindrome/description/

class Solution:
    def isPalindrome(self, s: str) -> bool:
        left, right = 0, len(s) - 1
        while left < right:
            # skip non-alphanumeric characters
            if not s[left].isalnum():
                left += 1
            elif not s[right].isalnum():
                right -= 1
            # compare the characters
            elif s[left].lower() != s[right].lower():
                return False
            else:
                left += 1
                right -= 1
        return True